File: web/src-server/query.cpp
/**
 * Web Server of the Chess Puzzle Database Explorer
 * Copyright (C) 2025  Lluís Alemany Puig
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 * Contact:
 *
 * 		Lluís Alemany Puig
 * 		https://github.com/lluisalemanypuig
 */

// C++ includes
#include "cpb/position.hpp"
#if defined DEBUG
#include <cassert>
#endif
#include <string_view>
#include <iostream>
#include <charconv>
#include <optional>

// cpb includes
#include <cpb/query.hpp>

static constexpr auto end = std::string::npos;

std::optional<int> to_int(const std::string_view s) noexcept
{
	int v;
	const auto [ptr, ec] = std::from_chars(s.begin(), s.end(), v);
	return ec == std::errc{} ? v : std::optional<int>{};
}

std::tuple<char, int, int> process_subfield(const std::string_view sub)
{
	const std::size_t colon = sub.find(':');
	if (colon == end) {
		std::cerr << __PRETTY_FUNCTION__ << '\n';
		std::cerr << "Could not find ':' in '" << sub << "'\n";
		return {};
	}
	const std::size_t comma = sub.find(',', colon);
	if (comma == end) {
		std::cerr << __PRETTY_FUNCTION__ << '\n';
		std::cerr << "Could not find ';' in '" << sub << "'\n";
		return {};
	}

	const std::string_view color{&sub[0], &sub[colon]};
	const std::string_view lb{&sub[colon + 1], &sub[comma]};
	const std::string_view ub{&sub[comma + 1], sub.end()};

	const std::optional<int> lb_int = to_int(lb);
	const std::optional<int> ub_int = to_int(ub);
	if (not lb_int or not ub_int) {
		std::cerr << __PRETTY_FUNCTION__ << '\n';
		std::cerr << "Missing valid numeric value in '" << sub << "'\n";
		std::cerr << "    lb_int: ";
		if (lb_int) {
			std::cerr << *lb_int << '\n';
		}
		else {
			std::cerr << "{}" << '\n';
		}
		std::cerr << "    ub_int: ";
		if (ub_int) {
			std::cerr << *ub_int << '\n';
		}
		else {
			std::cerr << "{}" << '\n';
		}
		return {};
	}

	return {color[0], *lb_int, *ub_int};
}

void process_piece_field(const std::string_view content, cpb::query_data& q)
{
	std::size_t pos = 0;
	std::size_t i = content.find(';', pos);

	while (i != end) {
		const std::string_view sub{&content[pos], &content[i]};
		const auto [color, lb_int, ub_int] = process_subfield(sub);
		if (color == 'w') {
			q.query_white = {.lb = lb_int, .ub = ub_int};
		}
		else if (color == 'b') {
			q.query_black = {.lb = lb_int, .ub = ub_int};
		}
		else if (color == 't') {
			q.query_both = {.lb = lb_int, .ub = ub_int};
		}

		pos = i + 1;
		i = content.find(';', pos);
	}
}

void parse_query_body(
	[[maybe_unused]] const std::string_view s, cpb::querier& Q
) noexcept
{
	Q.pawns.reset();
	Q.rooks.reset();
	Q.knights.reset();
	Q.bishops.reset();
	Q.queens.reset();
	Q.query_total_pieces = {};
	Q.query_player_turn = {};

	const std::size_t n_fields =
		static_cast<std::size_t>(std::count(s.begin(), s.end(), '['));

	std::size_t p = 0;
	for (std::size_t i = 0; i < n_fields; ++i) {
		const auto first = s.find('[', p);
		const auto second = s.find(']', p);
		if (first == end or second == end) {
			std::cerr << "Delimiter '[' or ']' could not be found.\n";
			std::cerr << "    Found [? " << (first == end ? "no\n" : "yes\n");
			std::cerr << "    Found ]? " << (second == end ? "no\n" : "yes\n");
			break;
		}

		const std::string_view name{&s[p], &s[first]};
		const std::string_view content{&s[first + 1], &s[second]};

		if (name == "p") {
			process_piece_field(content, Q.pawns);
		}
		else if (name == "r") {
			process_piece_field(content, Q.rooks);
		}
		else if (name == "k") {
			process_piece_field(content, Q.knights);
		}
		else if (name == "b") {
			process_piece_field(content, Q.bishops);
		}
		else if (name == "q") {
			process_piece_field(content, Q.queens);
		}
		else if (name == "T") {
			const auto [id, lb_int, ub_int] = process_subfield(content);
			if (id != 'T') {
				std::cerr << __PRETTY_FUNCTION__ << '\n';
				std::cerr << "Invalid identifier for total pieces: '" << id
						  << "'\n";
				break;
			}
			Q.query_total_pieces = {lb_int, ub_int};
		}
		else if (name == "M") {
			if (content != "w" and content != "b") {
				std::cerr << __PRETTY_FUNCTION__ << '\n';
				std::cerr << "Invalid turn indicator '" << content << "'\n";
				break;
			}
			Q.query_player_turn =
				content == "w" ? cpb::TURN_WHITE : cpb::TURN_BLACK;
		}
		else {
			std::cerr << "Invalid field indicator '" << name << "'.\n";
		}

		p = second + 1;
	}
}

File: web/src-client/query.ts
/**
 * Web Server of the Chess Puzzle Database Explorer
 * Copyright (C) 2025  Lluís Alemany Puig
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 * Contact:
 *
 * 		Lluís Alemany Puig
 * 		https://github.com/lluisalemanypuig
 */

export function query_elements(i: number, j: number, k: number): any | undefined {
	const white = document.getElementById(i.toString()) as HTMLInputElement;
	let white_bounds: any | undefined = undefined;
	if (white.checked) {
		const lower_id = i.toString() + "_2";
		const upper_id = i.toString() + "_3";
		const lower = document.getElementById(lower_id) as HTMLInputElement;
		const upper = document.getElementById(upper_id) as HTMLInputElement;
		white_bounds = {lower: lower.value, upper: upper.value};
	}

	const black = document.getElementById(j.toString()) as HTMLInputElement;
	let black_bounds: any | undefined = undefined;
	if (black.checked) {
		const lower_id = j.toString() + "_2";
		const upper_id = j.toString() + "_3";
		const lower = document.getElementById(lower_id) as HTMLInputElement;
		const upper = document.getElementById(upper_id) as HTMLInputElement;
		black_bounds = {lower: lower.value, upper: upper.value};
	}

	const total = document.getElementById(k.toString()) as HTMLInputElement;
	let total_bounds: any | undefined = undefined;
	if (total.checked) {
		const lower_id = k.toString() + "_2";
		const upper_id = k.toString() + "_3";
		const lower = document.getElementById(lower_id) as HTMLInputElement;
		const upper = document.getElementById(upper_id) as HTMLInputElement;
		total_bounds = {lower: lower.value, upper: upper.value};
	}
	return {white: white_bounds, black: black_bounds, total: total_bounds};
}

export function bounds_to_string(name: string, bounds: any): string {
	let str = "";
	let some_defined = false;
	if (bounds.white !== undefined) {
		str += `w:${bounds.white.lower},${bounds.white.upper};`;
		some_defined = true;
	}
	if (bounds.black !== undefined) {
		str += `b:${bounds.black.lower},${bounds.black.upper};`;
		some_defined = true;
	}
	if (bounds.total !== undefined) {
		str += `t:${bounds.total.lower},${bounds.total.upper};`;
		some_defined = true;
	}
	if (some_defined) {
		str = `${name}[${str}]`;
	}
	return str;
}

export function query_total_pieces(i: number): any | undefined {
	const total_pieces = document.getElementById(i.toString()) as HTMLInputElement;
	let total_bounds: any | undefined = undefined;
	if (total_pieces.checked) {
		const lower_id = i.toString() + "_2";
		const upper_id = i.toString() + "_3";
		const lower = document.getElementById(lower_id) as HTMLInputElement;
		const upper = document.getElementById(upper_id) as HTMLInputElement;
		total_bounds = {lower: lower.value, upper: upper.value};
	}
	return total_bounds;
}

export function total_pieces_to_string(name: string, total_pieces: any | undefined): string {
	if (total_pieces === undefined) {
		return "";
	}
	return `${name}[T:${total_pieces.lower},${total_pieces.upper}]`;
}

export function query_player_turn(i: number): string | undefined {
	const player_turn = document.getElementById(i.toString()) as HTMLInputElement;
	if (player_turn.checked) {
		const white_id = i.toString() + "_2";
		const black_id = i.toString() + "_3";
		const white = document.getElementById(white_id.toString()) as HTMLInputElement;
		const black = document.getElementById(black_id.toString()) as HTMLInputElement;
		if (white.checked) {
			return "w";
		}
		if (black.checked) {
			return "b";
		}
	}
	return undefined;
}

export function player_turn_to_string(name: string, player_turn: string | undefined): string {
	if (player_turn === undefined) {
		return "";
	}
	return `${name}[${player_turn}]`;
}
